hbase-endpoint/src/main/java/org/apache/hadoop/hbase/client/coprocessor/AsyncAggregationClient.java

```diff
@@ -131,32 +175,38 @@ private static byte[] nullToEmpty(byte[] b) {
     CompletableFuture<R> future = new CompletableFuture<>();
     AggregateRequest req;
     try {
-      req = validateArgAndGetPB(scan, ci, false);
+      req = validateArgAndGetPB(scan, ci, false, true);
     } catch (IOException e) {
       future.completeExceptionally(e);
       return future;
     }
-    AbstractAggregationCallback<R> callback = new AbstractAggregationCallback<R>(future) {
+    AbstractAggregationCallback<R> callback =
+      new AbstractAggregationCallback<>(future, req, AggregateService::getMax) {
 
-      private R max;
+        private R max;
+        private final Object lock = new Object();
```

Comment:
I'm adding concurrency safety to the aggregation callbacks. The aggregate() methods can be called by multiple threads at the same time. getFinalResult() in theory could be called by yet a different thread. The aggregation operations themselves need their critical sections synchronized, and the last aggregate() must happen-before getFinalResult().

Reply:
if multiple threads need to coordinate on retrieving this value, should they be synchronising on a final result? As in, should `getFinalResult` block on completion of a future Future instead of simply locking around accesses of `max` ?

Reply:
I'm playing with this idea. To do the accumulation of `max` I think you would need something like:
```
        private AtomicReference<CompletableFuture<R>> max;

        @Override
        protected void aggregate(RegionInfo region, AggregateResponse resp) throws IOException {
          if (resp.getFirstPartCount() > 0) {
            R result = getCellValueFromProto(ci, resp, 0);
            max.compareAndExchange(max, max.thenApply(m -> {
              if (m == null || (result != null && ci.compare(m, result) < 0)) {
                return result;
              } else {
                return m;
              }
            }));
          }
        }
```
except `AtomicReference::compareAndExchange` doesn't quite work the way I would need. Even if I could figure this out, I'm not sure this would be an improvement.

Reply:
Thanks for trying.